plutus-wallet-api/ledger/Ledger/Scripts.hs

```diff
@@ -229,6 +267,46 @@ to the previous problem: apply a function which does a pattern match and returns
 otherwise. Then, as before, we just check for error in the overall evaluation.
 -}
 
+{- Note [Sealing data scripts]
+To pass data scripts securely through the redeemer, we need to *seal* them. This means applying the 'seal' builtin
+to the data script value, and then applying the redeemer to all the (sealed) values.
+
+However, 'seal' is *polymorphic*, so we need to instantiate it at the correct type first. So we need to know the
+type of all the data scripts, which means we need to typecheck them. This is annoying, but I don't have
+another solution at the moment.
+-}
+
+{- Note [Impredicative polymorphism and seal]
+We need to turn 'seal :: forall a . a -> Sealed a' into a PLC term, which we
+can instantiate at a PLC type (which we do *not* have the corresponding Haskell type
+for). But this causes GHC to get its knickers in a twist in two ways:
+- GHC is very eager to replace the quantified type variable with 'Any', which we can't
+currently reconstruct. See Note [Polymorphic values and Any] for more complaining.
+- If you try and specify the type you end up asking for a 'CompiledCode (forall a . a -> Sealed a)'.
```

Comment:
I have no idea what I'm talking about, but can we have something like `forall a. CompiledCode (a -> Sealed a)`? Or can we instead of defining an unwrapper, provide a function that unwraps a `Sealer` and immediately instantiates it at a particular type? Again, I have no idea what I'm talking about.

Reply:
- `forall a . CompiledCode (a -> Sealed a)` has the wrong type. The point is for the type argument to the `CompiledCode` to be the same as the type of the program inside it. That program is definitely a type abstraction, so it really should be a quantified type if anything.
- The second approach is what I allude to in the second paragraph. The question is, what would the type of this function be? It would also be polymorphic! So we'd just have the same problem again.
    - Alternatively, we could try and generate it at runtime rather than using the compile-time plugin stuff. The main issue with this is that it has to interact with the wrapping/unwrapping of the `Sealer` type, which is difficult to to do with the runtime machinery. (Except the point of the hack is to realise that since it's a newtype it's *not* difficult to unwrap, so in a sense that's what we do, it's just that the function is the identity and we elide it.)

Reply:
> The point is for the type argument to the `CompiledCode` to be the same as the type of the program inside it.

I see, thanks for the explanation. Please bear with my nonsense a little longer: what if we fake the quantifier? Something like this:

```haskell
data Sealed a b where
    Sealed :: a -> Sealed a a

withUnsealed :: Sealed a b -> (a ~ b => a -> c) -> c
withUnsealed (Sealed x) k = k x
```

and then we can have

```haskell
newtype Sealer a = Sealer (forall b. b -> Sealed a b)
type CompiledSealed = forall a. CompiledCode (Sealer a)
```

This way we factor polymorphism out of `CompiledCode` and can later instantiate the `a` variable to whatever concrete type we need. And we still have that quantification to match with the PLC type. Just another random guess.

> The second approach is what I allude to in the second paragraph. The question is, what would the type of this function be? It would also be polymorphic! So we'd just have the same problem again.

What I attempted to communicate is whether we can have

```haskell
instSealerCode :: CompiledCode Sealer -> CompiledCode (a -> Sealed a)
```

as an additional primitive (where there is some connection between `a` and the PLC type that we instantiate the sealer at. I.e. there has to be a constraint or an additional argument or whatever)?

So instead of trying to solve the problem using the given interface, can we extend the interface?

Reply:
> what if we fake the quantifier? Something like this:

a) We don't have GADTs.
b) I think we still have a problem, which is unwrapping the `Sealer` newtype. `Sealer` is polymorphic, so a function to unwrap it will be polymorphic, and so we need to create a `CompiledCode <type of unwrap function>` to apply it... but then we're back where we started, I think.

> What I attempted to communicate is whether we can have ... as an additional primitive

I don't think making that work as a primitive is any easier than just making `seal :: forall a . a -> Sealed a` work as a primitive! 

Note that this all gets a lot easier once these *are* builtins, because then we can just have a reference to the builtin name. It's going via the plugin that makes this trick.

Reply:
> I don't think making that work as a primitive is any easier than just making `seal :: forall a . a -> Sealed` a work as a primitive!

Sorry, I mean just providing an additional function that is like `applyCode`, but only deals with instantiating `Sealer` at a particular type. That doesn't make sense?

Reply:
> Sorry, I mean just providing an additional function that is like applyCode, but only deals with instantiating Sealer at a particular type.

That's essentially what we're trying to write in the first place! The problem is making the piece of that function that actually unwraps the `Sealer` and applies it. We'd naturally want to do that with the plugin, since we're working with our encoding of Haskell types, but then all the same problems come up again.

Reply:
Given that we won't have this problem when it's a builtin, which is the plan anyway, I think we probably don't need to worry about it too much.